### PR TITLE
[REEF-413] Make the NameClient injectable

### PR DESCRIPTION
**Pull request by taegeonum**
This pull request addressed the issue by
 * Creates an interface of NameClient
 * Make the DefaultNameClientImpl injectable

JIRA: [REEF-413](https://issues.apache.org/jira/browse/REEF-413)

### CODE REVIEW

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java [hunk 1]
```diff
@@ -113,7 +113,7 @@
   @Inject
   public GroupCommDriverImpl(final ConfigurationSerializer confSerializer,
                              @Parameter(DriverIdentifier.class) final String driverId,
-                             @Parameter(TreeTopologyFanOut.class) final int fanOut) {
+                             @Parameter(TreeTopologyFanOut.class) final int fanOut) throws InjectionException {
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java]: Please don't expose the `Exception` here, as it changes the signature of the constructor, which breaks current users. Instead, we'll have to catch that exception and rethrow it as a `RuntimeException`. Not idea, but better than breaking users. Also, this is temporary anyhow, right?

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java [hunk 2]
```diff
@@ -155,7 +155,18 @@ public GroupCommDriverImpl(final ConfigurationSerializer confSerializer,
         groupCommFailedEvaluatorHandler);
     this.groupCommMessageHandler = new GroupCommMessageHandler();
     this.groupCommMessageStage = new SingleThreadStage<>("GroupCommMessageStage", groupCommMessageHandler, 100 * 1000);
-    this.netService = new NetworkService<>(idFac, 0, nameServiceAddr, nameServicePort,
+
+    Configuration nameClientConf = Tang.Factory.getTang().newConfigurationBuilder(NameClientConfiguration.CONF
+        .set(NameClientConfiguration.NAME_SERVER_HOSTNAME, nameServiceAddr)
+        .set(NameClientConfiguration.NAME_SERVICE_PORT, nameServicePort)
+        .build())
+        .bindImplementation(NameClient.class, DefaultNameClientImpl.class)
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java]: Do you need this? It should be the default anyhow, right?

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java [hunk 3]
```diff
@@ -155,7 +155,18 @@ public GroupCommDriverImpl(final ConfigurationSerializer confSerializer,
         groupCommFailedEvaluatorHandler);
     this.groupCommMessageHandler = new GroupCommMessageHandler();
     this.groupCommMessageStage = new SingleThreadStage<>("GroupCommMessageStage", groupCommMessageHandler, 100 * 1000);
-    this.netService = new NetworkService<>(idFac, 0, nameServiceAddr, nameServicePort,
+
+    Configuration nameClientConf = Tang.Factory.getTang().newConfigurationBuilder(NameClientConfiguration.CONF
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java]: `final`

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java [hunk 4]
```diff
@@ -155,7 +155,18 @@ public GroupCommDriverImpl(final ConfigurationSerializer confSerializer,
...
+        .set(NameClientConfiguration.NAME_SERVER_HOSTNAME, nameServiceAddr)
+        .set(NameClientConfiguration.NAME_SERVICE_PORT, nameServicePort)
+        .build())
+        .bindImplementation(NameClient.class, DefaultNameClientImpl.class)
+        .build();
+
+    Injector injector = Tang.Factory.getTang().newInjector(nameClientConf);
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java]: `final`

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java [hunk 5]
```diff
@@ -155,7 +155,18 @@ public GroupCommDriverImpl(final ConfigurationSerializer confSerializer,
...
+    NameClient nameClient = injector.getInstance(NameClient.class);
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java]: `final`

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java [hunk 6]
```diff
@@ -228,8 +239,9 @@ public Configuration getServiceConfiguration() {
             GroupCommNetworkHandlerImpl.class)
         .bindNamedParameter(NetworkServiceParameters.NetworkServiceExceptionHandler.class,
             ExceptionHandler.class)
-        .bindNamedParameter(NameServerParameters.NameServerAddr.class, nameServiceAddr)
-        .bindNamedParameter(NameServerParameters.NameServerPort.class, Integer.toString(nameServicePort))
+        .bindNamedParameter(NameClientParameters.NameServerAddr.class, nameServiceAddr)
+        .bindNamedParameter(NameClientParameters.NameServerPort.class, Integer.toString(nameServicePort))
+        .bindNamedParameter(NetworkServiceParameters.NetworkServiceNameClient.class, DefaultNameClientImpl.class)
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java]: Same as above: If it is the default, we won't need the binding.

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkService.java [hunk 7]
```diff
@@ -81,63 +61,17 @@
...
-   * @param nameServerAddr
-   * @param nameServerPort
-   * @param codec
-   * @param tpFactory
-   * @param recvHandler
-   * @param exHandler
-   * @deprecated have an instance injected instead.
-   */
-  @Deprecated
-  public NetworkService(final IdentifierFactory factory,
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkService.java]: Same as on the other PR: Please do not remove these old constructors. Instead, be super-explicit in the `@deprecated` message and file a JIRA conditioned on release 0.12 to delete these.

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkService.java [hunk 8]
```diff
@@ -81,63 +61,17 @@
...
-
-  /**
    * @deprecated have an instance injected instead.
    */
   @Deprecated
   public NetworkService(final IdentifierFactory factory,
                         final int nsPort,
-                        final String nameServerAddr,
-                        final int nameServerPort,
+                        final NameClient nameClient,
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkService.java]: We still need the old constructor, without that parameter ...

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkService.java [hunk 9]
```diff
@@ -148,15 +82,12 @@ public NetworkService(
...
-      @Parameter(NameServerParameters.NameServerPort.class) final int nameServerPort,
-      @Parameter(NameLookupClient.RetryCount.class) final int retryCount,
-      @Parameter(NameLookupClient.RetryTimeout.class) final int retryTimeout,
+      @Parameter(NetworkServiceParameters.NetworkServiceNameClient.class) final NameClient nameClient,
       @Parameter(NetworkServiceParameters.NetworkServiceCodec.class) final Codec<T> codec,
       @Parameter(NetworkServiceParameters.NetworkServiceTransportFactory.class) final TransportFactory tpFactory,
       @Parameter(NetworkServiceParameters.NetworkServiceHandler.class) final EventHandler<Message<T>> recvHandler,
       @Parameter(NetworkServiceParameters.NetworkServiceExceptionHandler.class) final EventHandler<Exception> exHandler,
-      final LocalAddressProvider localAddressProvider) {
+      final LocalAddressProvider localAddressProvider) throws InjectionException {
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkService.java]: Please don't throw, as it breaks current users.

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkServiceParameters.java [hunk 10]
```diff
@@ -58,4 +59,7 @@
   public static class NetworkServiceExceptionHandler implements Name<EventHandler<?>> {
   }
 
+  @NamedParameter(doc = "name client")
+  public static class NetworkServiceNameClient implements Name<NameClient> {
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkServiceParameters.java]: Why would we need a `Name` for the `NameClient`? Isn't it a matter of binding an implementation?

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClient.java [hunk 11]
```diff
@@ -19,257 +19,11 @@
...
-    if (value instanceof NamingLookupResponse) {
-      replyLookupQueue.offer((NamingLookupResponse) value);
-    } else if (value instanceof NamingRegisterResponse) {
-      replyRegisterQueue.offer((NamingRegisterResponse) value);
-    } else {
-      throw new NamingRuntimeException("Unknown naming response message");
-    }
-
-  }
+public interface NameClient extends Stage, Naming {
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClient.java]: Add a `@DefaultImplementation` annotation.

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClientConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  /**
+   * The port used by name server.
+   */
+  public static final RequiredParameter<Integer> NAME_SERVICE_PORT = new RequiredParameter<>();
+  /**
+   * DNS hostname running the name service.
+   */
+  public static final RequiredParameter<String> NAME_SERVER_HOSTNAME = new RequiredParameter<>();
+
+  public static final OptionalParameter<IdentifierFactory> IDENTIFIER_FACTORY = new OptionalParameter<>();
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClientConfiguration.java]: Please document all these.

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClientParameters.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.apache.reef.io.network.util.StringIdentifierFactory;
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+import org.apache.reef.wake.IdentifierFactory;
+
+/**
+ *
+ */
+public final class NameClientParameters {
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClientParameters.java]: Can you please move these to be outer classes in a package `parameters`? That way, C# and Java can eventually use the same names for the named parameters. (C# encodes inner classes different from Java).

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClientParameters.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+ */
+package org.apache.reef.io.network.naming;
+
+import org.apache.reef.io.network.util.StringIdentifierFactory;
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+import org.apache.reef.wake.IdentifierFactory;
+
+/**
+ *
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClientParameters.java]: Please don't make empty JavaDocs. It fools tools into thinking we had documentation where we have none :)

**Comment by markusweimer** [comment]
@taegeonum I did a pass. Similar to #247, we need to be careful not to break current users with this.

**Comment by taegeonum** [comment]
@markusweimer is it ok to change NameClient implementation to DefaultNameClientImipl and make NameClient interface?

**Comment by taegeonum** [comment]
@markusweimer Thanks for your comments. I've addressed your comments. Please take a look at them.

**Comment by markusweimer** [comment]
The move of `NameClient` from class to interface does indeed break code, as `new NameClient(...)` no longer applies. Maybe we should keep it as a class for now, and then make the move to an interface once we are sure people only get the instance via Tang. Does that sound feasible?

**Comment by taegeonum** [comment]
@markusweimer I'm thinking about two different implementations of NameClient. First is DefaultNameClientImpl which lookups a remote NameServer, and Second is LocalLookupNameClientImpl which lookups a local NameServer. The interface of `NameClient` is needed in order to inject different implementation.

**Comment by markusweimer** [comment]
This is tricky, as changing a class to an interface creates an incompatible API. We could introduce *another* name of the interface. How about `NameResolver`? That way, the old `NameClient` can maintain its name and implement `NameResolver` in 0.12, and it can be renamed to `RemoteNameResolver` in 0.13. What do you think?

Also: There are now merge conflicts on this PR, please resolve or maybe even rebase.

**Comment by taegeonum** [comment]
@markusweimer Sounds good idea! I've created the `NameResolver`. Please take a look

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClientConfiguration.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.reef.tang.formats.ConfigurationModule;
+import org.apache.reef.tang.formats.ConfigurationModuleBuilder;
+import org.apache.reef.tang.formats.OptionalParameter;
+import org.apache.reef.tang.formats.RequiredParameter;
+import org.apache.reef.wake.IdentifierFactory;
+
+/**
+ * Configuration Module Builder for NameClient.
+ */
+public final class NameClientConfiguration extends ConfigurationModuleBuilder {
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClientConfiguration.java]: This should be named `NameResolverConfiguration` now, right?

**Comment by markusweimer** [comment]
Almost there, just one more rename :-)

**Comment by taegeonum** [comment]
@markusweimer Done!

**Comment by markusweimer** [comment]
Great, I will test and merge.

**Pull request closed without merging**